File: java_solutions/374.guess-number-higher-or-lower.java
/*
 * @lc app=leetcode id=374 lang=java
 *
 * [374] Guess Number Higher or Lower
 */

public class Solution extends GuessGame {
    public int guessNumber(int n) {
        return Stream.iterate(new long[]{1, (long)n, 0, -2}, p -> Optional.of(p[0] + (p[1] - p[0]) / 2).map(m -> Optional.of((long)guess(m.intValue())).map(g -> new long[]{g == 1 ? m + 1 : p[0], g == -1 ? m - 1 : p[1], m, g}).get()).get()).filter(p -> p[3] == 0).findFirst().map(p -> (int)p[2]).get();
    }
}

File: kotlin_solutions/374.guess-number-higher-or-lower.kt
/*
 * @lc app=leetcode id=374 lang=kotlin
 *
 * [374] Guess Number Higher or Lower
 */

class Solution:GuessGame() { override fun guessNumber(n:Int):Int { return bs(1, n) } tailrec fun bs(left: Int, right: Int): Int { return (left + (right - left) / 2).let { mid -> when { guess(mid) == 0 -> {mid} guess(mid) == 1 -> {bs(mid + 1, right)} else -> bs(left, mid - 1) } } } }
